src/backend/distributed/planner/recursive_planning.c

```diff
@@ -0,0 +1,606 @@
+	{
+		Query *query = (Query *) node;
+		bool found = false;
+		int flags = 0;
+
+		context->level += 1;
+		found = query_tree_walker(query, ContainsReferencesToOuterQueryWalker,
+								  context, flags);
+		context->level -= 1;
+
+		return found;
+	}
+	else
+	{
+		return expression_tree_walker(node, ContainsReferencesToOuterQueryWalker,
+									  context);
+	}
+}
+
+
+/*
+ * BuildSubPlanResultQuery returns a query of the form:
+ *
+ * SELECT
+ *   <target list>
+ * FROM
+ *   read_intermediate_result('<planId>_<subPlanId>', '<copy format'>)
+ *   AS res (<column definition list>);
+ *
+ * The target list and column definition list are derived from the given subquery.
+ *
+ * If any of the types in the target list cannot be used in the binary copy format,
+ * then the copy format 'text' is used, otherwise 'binary' is used.
+ */
+static Query *
+BuildSubPlanResultQuery(Query *subquery, uint64 planId, uint32 subPlanId)
+{
+	Query *resultQuery = NULL;
+	char *resultIdString = NULL;
+	Const *resultIdConst = NULL;
+	Const *resultFormatConst = NULL;
+	FuncExpr *funcExpr = NULL;
+	Alias *funcAlias = NULL;
+	List *funcColNames = NIL;
+	List *funcColTypes = NIL;
+	List *funcColTypMods = NIL;
+	List *funcColCollations = NIL;
+	RangeTblFunction *rangeTableFunction = NULL;
+	RangeTblEntry *rangeTableEntry = NULL;
+	RangeTblRef *rangeTableRef = NULL;
+	FromExpr *joinTree = NULL;
+	ListCell *targetEntryCell = NULL;
+	List *targetList = NIL;
+	int columnNumber = 1;
+	bool useBinaryCopyFormat = true;
+	Oid copyFormatId = BinaryCopyFormatId();
+
+	/* build the target list and column definition list */
+	foreach(targetEntryCell, subquery->targetList)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) lfirst(targetEntryCell);
+		Node *targetExpr = (Node *) targetEntry->expr;
+		char *columnName = targetEntry->resname;
+		Oid columnType = exprType(targetExpr);
+		Oid columnTypMod = exprTypmod(targetExpr);
+		Oid columnCollation = exprCollation(targetExpr);
+		Var *functionColumnVar = NULL;
+		TargetEntry *newTargetEntry = makeNode(TargetEntry);
+
+		if (targetEntry->resjunk)
+		{
+			continue;
+		}
+
+		funcColNames = lappend(funcColNames, makeString(columnName));
+		funcColTypes = lappend_int(funcColTypes, columnType);
+		funcColTypMods = lappend_int(funcColTypMods, columnTypMod);
+		funcColCollations = lappend_int(funcColCollations, columnCollation);
+
+		functionColumnVar = makeNode(Var);
+		functionColumnVar->varno = 1;
+		functionColumnVar->varattno = columnNumber;
+		functionColumnVar->vartype = columnType;
+		functionColumnVar->vartypmod = columnTypMod;
+		functionColumnVar->varcollid = columnCollation;
+		functionColumnVar->varlevelsup = 0;
+		functionColumnVar->varnoold = 1;
+		functionColumnVar->varoattno = columnNumber;
+		functionColumnVar->location = -1;
+
+		newTargetEntry = makeNode(TargetEntry);
+		newTargetEntry->expr = (Expr *) functionColumnVar;
+		newTargetEntry->resno = columnNumber;
+		newTargetEntry->resname = columnName;
```

Comment:
Sure. There are some other fields in all these Nodes which are initialised to 0 by `makeNode`, but `resjunk` appears in our code common enough that it might make sense to show it here.